packages/storage/amplify_storage_s3/example/backend/lib/stack.ts

```diff
@@ -0,0 +1,309 @@
+      [StorageAccessLevel.public]: props.prefixResolver
+        ? props.prefixResolver(StorageAccessLevel.public, identityId)
+        : "public/",
+      [StorageAccessLevel.protected]: props.prefixResolver
+        ? props.prefixResolver(StorageAccessLevel.protected, identityId)
+        : `protected/${identityId}/`,
+      [StorageAccessLevel.private]: props.prefixResolver
+        ? props.prefixResolver(StorageAccessLevel.private, identityId)
+        : `private/${identityId}/`,
+    };
+    const buildPolicyDocument = (accessLevel: StorageAccessLevel) => {
+      let permissions = [
+        StoragePermission.get,
+        StoragePermission.put,
+        StoragePermission.delete,
+      ];
+      if (props.accessLevels) {
+        permissions = props.accessLevels[accessLevel];
+      }
+      const prefix = prefixes[accessLevel];
+      return new iam.PolicyDocument({
+        statements: [
+          new iam.PolicyStatement({
+            actions: permissions,
+            resources: [`${bucket.bucketArn}/${prefix}*`],
+          }),
+        ],
+      });
+    };
+
+    const unauthenticatedPolicies: Record<string, iam.PolicyDocument> = {
+      "read-unauth": new iam.PolicyDocument({
+        statements: [
+          new iam.PolicyStatement({
+            actions: ["s3:ListBucket"],
+            conditions: {
+              StringLike: {
+                "s3:prefix": [
+                  `${prefixes[StorageAccessLevel.public]}`,
+                  `${prefixes[StorageAccessLevel.public]}*`,
+                  `${prefixes[StorageAccessLevel.protected]}`,
+                  `${prefixes[StorageAccessLevel.protected]}*`,
+                ],
+              },
+            },
+            resources: [bucket.bucketArn],
+          }),
+        ],
+      }),
+      public: buildPolicyDocument(StorageAccessLevel.public),
+      protected: buildPolicyDocument(StorageAccessLevel.protected),
+    };
+    const unauthenticatedRole = new iam.Role(this, "UnauthenticatedRole", {
+      description: "Default role for anonymous users",
+      assumedBy: new iam.FederatedPrincipal(
+        "cognito-identity.amazonaws.com",
+        {
+          StringEquals: {
+            "cognito-identity.amazonaws.com:aud": identityPool.ref,
+          },
+          "ForAnyValue:StringLike": {
+            "cognito-identity.amazonaws.com:amr": "unauthenticated",
+          },
+        },
+        "sts:AssumeRoleWithWebIdentity"
+      ),
+    });
+
+    const authenticatedPolicies: Record<string, iam.PolicyDocument> = {
+      ...unauthenticatedPolicies,
+      "read-auth": new iam.PolicyDocument({
+        statements: [
+          new iam.PolicyStatement({
+            actions: ["s3:ListBucket"],
+            conditions: {
+              StringLike: {
+                "s3:prefix": [
+                  `${prefixes[StorageAccessLevel.private]}`,
+                  `${prefixes[StorageAccessLevel.private]}*`,
+                ],
```

Comment:
Oops, wasn't aware until actually tested... looks like we are missing read permission of `protected/` and `protected/*` here.

I noticed that the IAM policies created after deploy the stack are slightly different from the IAM policies created by Amplify CLI. It may be worth it to precisely replicate the Amplify CLI created IAM policy to ensure the tests results are accurate?

Reply:
Updated. I believe it should match now 👍

Reply:
TY for the quick fix, will continue adding integration tests based on this new change.